Invoke component cleanup in the Operator controller on Submariner deletion

tpantelis opened the pull request:
Create `DaemonSets`/`Deployments` to uninstall the **submariner-gateway**, **submariner-routeagent**, **submariner-globalnet**, and **submariner-networkplugin-syncer** components as outlined in #1722. See individual commits for details.

Fixes #1722

tpantelis committed:
Refactor submariner_controller_test

Add testDriver struct and other simplifications.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Add finalizer to Submariner in the operator controller

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Invoke uninstall for the gateway DaemonSet

Delete the regular gateway DaemonSet, create the
cleanup/uninstall DaemonSet, await readiness and delete it.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Invoke uninstall for the route-agent DaemonSet

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Invoke uninstall for the globalnet DaemonSet

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Invoke uninstall for the network plugin syncer Deployment

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Abort Submariner component cleanup if max time elapsed

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

Jaanki commented on controllers/submariner/submariner_controller_test.go:
```diff
@@ -20,78 +20,23 @@ package submariner_test
 const (
-	submarinerName          = "submariner-gateway"
+	submarinerName          = "submariner"
 	submarinerNamespace     = "submariner-operator"
```
These 2 constants are defined in `internal/constants/constants.go` as well. That could be used here.

tpantelis replied:
These predated this PR. But the controller code does not/should not rely on the global constants (the namespace at least) so neither should the test code. In fact we should use values that differ from the global constants to ensure the controller code does not use them. Changed with https://github.com/submariner-io/submariner-operator/pull/1803/commits/06112225d96942e0e0a214c4d494970209650b91.

Jaanki commented on controllers/submariner/cleanup.go:
```diff
@@ -39,14 +39,21 @@ import (
 		}
+	}
 
+	// Next, delete ensure all their pods are cleaned up.
```
// Next, delete `and` ensure all their pods are cleaned up.?

tpantelis replied:
yeah that's fixed in a subsequent commit.

Jaanki commented on controllers/submariner/submariner_controller_test.go:
```diff
@@ -35,9 +35,8 @@ import (
-	submarinerName          = "submariner"
-	submarinerNamespace     = "submariner-operator"
-	routeAgentDaemonSetName = "submariner-routeagent"
+	submarinerName      = "submariner"
```
reuse these from `internal/constants/constants.go`?

tpantelis replied:
Same as above.

tpantelis committed:
Change Submariner name/namespace in submariner_controller_test

...so they're different than the constants defined in
internal/constants/constants.go. This ensures that the controller
code does not reference them and assume hardcoded values.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Elide component uninstall if not supported by the version

The first release version for the uninstall functionality will
be 0.12.0 so check for that.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

dfarrell07 commented on controllers/submariner/submariner_controller_test.go:
```diff
@@ -239,14 +238,16 @@ func testDeletion() {
 	})
 
-	It("should uninstall the component resources and remove the finalizer", func() {
+	FIt("should uninstall the component resources and remove the finalizer", func() {
```
Typo? s/FIt/It

tpantelis replied:
F = focus - I removed the F in the next commit.

dfarrell07 replied:
Ah okay, thanks/learning :nerd_face: :+1:

dfarrell07 replied:
(I noticed it went away btw, that's part of why I thought it was a typo)

skitt commented on controllers/submariner/np_syncer_resources.go:
```diff
@@ -123,3 +123,7 @@ func newNetworkPluginSyncerDeployment(cr *v1alpha1.Submariner, clusterNetwork *n
 	return networkPluginSyncerDeployment
 }
+
+func canInstallNetworkPluginSyncer(submariner *v1alpha1.Submariner) bool {
```
Bit of a nitpick, but wouldn’t this be better as `needsNetworkPluginSyncer`?

tpantelis replied:
Sure

skitt commented on controllers/submariner/submariner_controller.go:
```diff
@@ -239,6 +246,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 			// TODO skitt determine what we should really be doing for concurrent updates to the Submariner CR
 			// Updates fail here because the instance is updated between the .Update() at the start of the function
 			// and the status update here
+			fmt.Printf("****UPDATE: %#v", err)
```
Is this a leftover?

tpantelis replied:
Crap - yes.

sridhargaddam commented on controllers/submariner/submariner_controller.go:
```diff
@@ -59,6 +61,7 @@ import (
 const (
 	gatewayMetricsServerPort   = 8080
 	globalnetMetricsServerPort = 8081
+	SubmarinerFinalizer        = "submariner-controller-finalizer"
```
nit: Usually, I have seen people using finalizers like 
```
    finalizers:                                                                                                                                                                              
    - managedcluster-import-controller.open-cluster-management.io/cleanup                                                                                                                    
    - managedclusterinfo.finalizers.open-cluster-management.io                                                                                                                               
    - cluster.open-cluster-management.io/api-resource-cleanup                                                                                                                                
    - agent.open-cluster-management.io/klusterletaddonconfig-cleanup                                                                                                                         
    - open-cluster-management.io/managedclusterrole                                                                                                                                          
    - managedcluster-import-controller.open-cluster-management.io/manifestwork-cleanup                                                                                                       
    - cluster.open-cluster-management.io/submariner-agent-cleanup
```

So shall we follow the same pattern and name it something like `submariner.io/cleanup" ?

tpantelis replied:
Done with https://github.com/submariner-io/submariner-operator/pull/1803/commits/31cfc9d3e374c0eae00bd6c06f5f2799668957f5

sridhargaddam commented on controllers/submariner/cleanup.go:
```diff
@@ -20,15 +20,148 @@ package submariner
+	if daemonSet.Status.DesiredNumberScheduled == 0 {
+		log.Info("DaemonSet has no available nodes:", "name", daemonSet.Name, "namespace", daemonSet.Namespace)
+	} else if daemonSet.Status.DesiredNumberScheduled != daemonSet.Status.NumberReady {
+		log.Info("DaemonSet not ready yet - requeueing:", "name", daemonSet.Name, "namespace", daemonSet.Namespace,
```
What would happen if the admin marks one of the clusterNodes as unschedulable (ie.., kubectl cordon node-name)? Will we continue to retry as the desiredNumber will not match numberReady?

tpantelis replied:
I don't know how K8s handles an unschedulable node, ie is it counted in `DesiredNumberScheduled` (logically I wouldn't think so)? Either way we will eventually timeout and delete the `DaemonSet`.

tpantelis replied:
According to [this](https://kubernetes.io/docs/concepts/architecture/nodes/), "_Marking a node as unschedulable prevents the scheduler from placing new pods onto that Node_" so i'd think `DesiredNumberScheduled` wouldn't reflect it.

sridhargaddam commented on controllers/submariner/cleanup.go:
```diff
@@ -20,15 +20,148 @@ package submariner
+	// See http://blog.itaysk.com/2017/12/26/the-single-use-daemonset-pattern-and-prepulling-images-in-kubernetes
+	// for more details.
+	podSpec.InitContainers = podSpec.Containers
+	podSpec.InitContainers[0].Env = append(podSpec.InitContainers[0].Env, corev1.EnvVar{Name: "UNINSTALL", Value: "true"})
```
So is the init container responsible to do the necessary cleanup on the respective node?
Where are we passing the script (or the code) to perform the cleanup?

tpantelis replied:
> So is the init container responsible to do the necessary cleanup on the respective node? 

Yes

> Where are we passing the script (or the code) to perform the cleanup?

I copy the normal container to the init container (L149) so it's the same image and params except I add the "UNINSTALL" env var.

sridhargaddam reviewed (commented):
Thanks for the PR @tpantelis. Its a jumbo PR but the individual commits helped.

This PR/feature deserves a tech-exchange session ;-)

tpantelis committed:
Rename controller finalizer

"submariner-controller-finalizer" ->
"controllers.submariner.io/cleanup".

Also renamed the constant to CleanupFinalizer and moved to a
common file so it can be used by the service disc controller.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

sridhargaddam reviewed (approved):
Thanks @tpantelis

skitt committed:
Merge branch 'devel' into operator_uninstall
